cmd/podman/generate/spec.go

```diff
@@ -0,0 +1,69 @@
+package pods
+
+import (
+	"fmt"
+
+	"github.com/containers/podman/v4/cmd/podman/common"
+	"github.com/containers/podman/v4/cmd/podman/registry"
+	"github.com/containers/podman/v4/cmd/podman/utils"
+	"github.com/containers/podman/v4/pkg/domain/entities"
+	"github.com/spf13/cobra"
+)
+
+var (
+	specCmd = &cobra.Command{
+		Use:               "spec [options] {CONTAINER... |POD...}",
+		Short:             "Generate Specgen JSON based on containers or pods",
+		Long:              "Generate Specgen JSON based on containers or pods",
+		RunE:              spec,
+		Args:              cobra.ExactArgs(1),
+		ValidArgsFunction: common.AutocompleteContainers,
+		Example:           `podman generate spec ctrID`,
+	}
+)
+
+var (
+	opts *entities.GenerateSpecOptions
+)
+
+func init() {
+	registry.Commands = append(registry.Commands, registry.CliCommand{
+		Command: specCmd,
+		Parent:  generateCmd,
+	})
+	opts = &entities.GenerateSpecOptions{}
+	flags := specCmd.Flags()
+
+	filenameFlagName := "filename"
+	flags.StringVarP(&opts.FileName, filenameFlagName, "f", "", "Write output to the specified path")
+
+	compactFlagName := "compact"
+	flags.BoolVarP(&opts.Compact, compactFlagName, "c", false, "Print the json in a compact format for consumption")
+
+	nameFlagName := "name"
+	flags.StringVarP(&opts.Name, nameFlagName, "n", "", "Specify a new name for the generated spec")
+
+	flags.SetNormalizeFunc(utils.AliasFlags)
+}
+
+func spec(cmd *cobra.Command, args []string) error {
+	opts.ID = args[0]
+
+	if len(opts.Name) == 0 {
+		return fmt.Errorf("must specify a new name for the container/pod")
```

Comment:
I don't like the name requirement as the user faces the challenge of finding a unique name.  I think that's Podman's job.

Without having looked at the details, my intuitive expectation is that Podman would be clever enough to assign a name if none is set.  That would make such spec-files _portable_.

Similar to `generate systemd --new`, I imagine a "generate once deploy as often as you want" kind of workflow.  Enforcing a name choise would complicate the workflow quite a bit.

Somehow related: mandatory options/flags should be arguments.

Reply:
What does `podman generate kube` do? We should follow that precedent.

Reply:
`generate kube` does not alter the name(s) of pods/containers.  `kube play` will error out though when trying to create a pod/container with a conflicting name.

Reply:
So shouldn't this work the same, and force users to change the name?  IE If we are going to allow users to modify and ship the spec?

Reply:
I think that sounds like a sane default.

I also see beauty in not setting the name at all in the spec.  This way, the files are really portable as podman can create a unique name without the risk of a name conflict when the file is deployed on another machine.  I do not have a strong opinion whether clearing the name should be the default or not.  Probably worth adding a flag?

* `podman generate spec --name=false` -> clears the name
* `podman generate spec --name=true` -> generates with the name

But I am focusing on containers only.  I am not sure whether clearing names would work for pods.

In case the feature gets promoted for higher tasks we should check how we want it to work in automated fashion.

Reply:
I have a function used by clone that can generate a new name for both pods and containers. I will add the `--name=t/f` flag I think that makes the most sense.

it is basically the same as my `--clone` flag from a previous commit just with a different, more general name.

Reply:
What's important IMO is that we can generate a file once and deploy it anywhere without risking a name conflict on a node.  As long as Podman can auto-assign a name when _reading_/_running_ the spec, I am happy :)

Reply:
The crux of the matter is to assign names when applying/running/playing the file and not when generating it.

Reply:
ok so @vrothberg @rhatdan, here an auto-generation might be warranted, but when if/when we implement a `run --spec` we can offer a flag to rename?